### json 불러와서 저장하기

In [1]:
import numpy as np
import pandas as pd
import json
import jpype
import glob
from random import shuffle
from bs4 import BeautifulSoup as bs

data_path = "data"
#data_path = "data(test)"

file_list = glob.glob("%s/*.json" % data_path)

shuffle(file_list)

json_train=[]

for json_file_name in file_list:
    json_file = json.loads(open(json_file_name).read())
    json_train += json_file["articles"]

### json에서 title 형태소 추출

In [2]:
from konlpy.tag import Mecab

mecab = Mecab()

labeled_train = []

for cnt, article in enumerate(json_train):
    
    if cnt % 10000 == 0:
        print(cnt)
        
    title_pos = ["%s_%s" % (word, pos) for word, pos in mecab.pos(article["title"])]

    labeled_train.append({
            "istroll": article["is_troll"],
            "title_pos": title_pos,
            "title_pos_sentences" : " ".join(title_pos)
        })

labeled_train = pd.DataFrame.from_dict(labeled_train)
labeled_train

0
10000
20000
30000
40000
50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000


istroll                                          title_pos  \
0        False  [명절_NNG, 이_VCP, 지만_EC, 참_VV, 을_ETM, 수_NNB, 없_V...   
1        False  [이젠_NNG+JX, 아니_VCN, 라는_ETM, 증거_NNG, 를_JKO, 대봐_...   
2        False      [뻘_NNG, 글_NNG, +_SY, 오늘_NNG, 의_JKG, 일러스트_NNG]   
3        False                 [일_NR, 일_NNBC, 단식_NNG, 보_VV, 고_EC]   
4        False                            [간헐_NNG, 적_XSN, 단식_NNG]   
5        False                      [옥상_NNG, 에서_JKB, ㅇㅅㅇ_UNKNOWN]   
6        False  [강아지_NNG, 정수기_NNG, 를_JKO, 알아보_VV, 고_EC, 있_VX, ...   
7        False                  [어_IC, 그_NP, 로_JKB, 성공_NNG, !_SF]   
8        False                             [웹툰_NNP, -_SY, 터널_NNG]   
9        False               [혜리_NNG, 의_JKG, 질투_NNG, ㅋㅋㅋ_UNKNOWN]   
10       False  [[_SSO, 꾸준_XR, 글_NNG, ]_SSC, 안녕_NNG, 하_XSV, 세요...   
11       False         [으앙_MAG, 밤_NNG, 이_VCP, 군요_EF, ㅇㅅㅇ_UNKNOWN]   
12       False  [벌레_NNG, 들_XSN, 지_NNG, 들_XSN, 폐쇄_NNG, 서명_NNG, ...   
13       False  [여기_NP, 서_JKB, 농_NNG, 닭_NNG, 체_NNB, 쓰_VV, 다가_E...   
14       False  [이거_NP, 왜_MAG, 농약_NNG, 글_NNG, 이_VCP, 에요_EF, ?_SF]   
15       False  [2_SN, 년_NNB, 째_XSN, 명절_NNG, 에_JKB, 고향_NNG, 에_...   
16       False  [여왕_NNG, 폐하_NNG, 가_JKS, 일_NR, 베_NNG, 유저_NNG, 에...   
17       False  [벌레_NNG, 들_XSN, 이_JKS, 어디_NP, 에서_JKB, 사_VV, 는지...   
18       False                       [신기_NNG, 방기_NNG, ._SF, ._SF]   
19       False                [맥주_NNG, 두_MM, 캔_NNBC, 째_XSN, ._SF]   
20       False  [소고기_NNG, 는_JX, 미국_NNP, 산_XSN, 소고기_NNG, 가_JKS,...   
21       False                 [Free_SL, tempo_SL, -_SY, rain_SL]   
22       False  [저번_NNG, 에_JKB, 광화문_NNG, 에서_JKB, 햄버거_NNG, 먹_VV...   
23       False  [[_SSO, 펌_NNG, ]_SSC, 약자_NNG, 가_JKS, 약자_NNG, 를...   
24       False  [새_MM, 누리_NNG, 의_JKG, '_SY, 특검_NNG, 한계_NNG, 인정...   
25       False     [오늘_MAG, 귀농_NNG, 했_XSV+EP, 습니_EC, 디_NNG, !_SF]   
26       False                      [얼굴_NNG, 만_JX, 반쪽_NNG, 이_JKS]   
27       False  [리버_NNP, 오피스_NNP, 실무_NNG, 에서_JKB, 사용_NNG, 중_NN...   
28       False  [세월_NNG, 호_NNBC, 충돌_NNG, 의혹_NNG, 과_JC, 레이더_NNG...   
29       False          [강원도_NNP, ._SF, .._SY, 흠_IC, ._SF, .._SY]   
...        ...                                                ...   
172014   False                             [배고프_VA, 다아아_EF, 앍_VV]   
172015    True  [CNN_SL, ,_SC, 무인_NNG, 항공기_NNG, 장난감_NNG, 가_VV,...   
172016   False  [윤태호_NNP, 작가_NNP, 님_XSN, 의_JKG, 내부자_NNG, 들_XSN...   
172017   False     [정권_NNG, 별_XSN, 군사비_NNG, 비교_NNG, ._SY, jpg_SL]   
172018   False                   [그거_NP, 생각난다_VV+EC, 농_NNG, ._SF]   
172019    True  [농민_NNG, 님_XSN, 들_XSN, 새로_MAG, 가입_NNG, 했_XSV+E...   
172020   False  [밀_VV+ETM, 덕_NNG, 이_VCP, 라면_EC, 알_VV, 겠_EP, 지만...   
172021    True  [왜_MAG, 솜방망이_NNG, 처벌_NNG, 이_JKS, 계속_MAG, 되_VV,...   
172022   False  [뱃살_NNG, 이_JKS, 찌_VV, 니까_EC, 한_MM, 의원_NNG, 광고_...   
172023   False         [일간_NNG, 워스트_NNP, 시작_NNG, 했_XSV+EP, 닭_NNG]   
172024   False  [동성애자_NNG, 말_VV, 고_EC, 다른_MM, 성소_NNG, 수자_NNG, ...   
172025    True  [김새론_NNP, 10_SN, 몇_NR, 년_NNB, 사이_NNG, 에_JKB, 많...   
172026   False          [다_MAG, 끄_VV, 네_EF, 식_NNB, 대화법_NNG, ._SF]   
172027   False                                        [악동뮤지션_NNP]   
172028   False  [이것_NP, 도_JX, 엄밀히_MAG, 말_NNG, 하_XSV, 면_EC, 광고_...   
172029   False  [[_SSO, 일_NNG, 베_VV+EC, 스_VV, 샷_EP+EP, ]_SSC, ...   
172030   False  [부부_NNG, 의_JKG, 대화_NNG, (_SSO, 말꼬리_NNG, 에_JKB,...   
172031   False  [오랜만_NNG, 에_JKB, 와_VV+EC, 봤_VX+EP, 는데_EC, 잡담_N...   
172032   False  [남자_NNG, 들_XSN, 이_JKS, 버스_NNG, 탔_VV+EP, 을_ETM,...   
172033   False                      [술_NNG, 꾼_XSN, 의_JKG, 결심_NNG]   
172034   False  [닝_NNP, 겐_JKB+JX, ,_SC, 내_NP, 가_JKS, 시범_NNG, 을...   
172035   False  [동성애자_NNG, 에_JKB, 대한_VV+ETM, 박_NNP, 대통령_NNG, 의...   
172036    True  [역시_MAJ, 일_NR, 워_NNG, 가_JKS, 망할_VV+ETM, 줄_NNB,...   
172037    True                        [유익_XR, 한_XSA+ETM, 사이트_NNG]   
172038   False  [이게_NP+JKS, 뭔_MM, 개소리_NNG, 농_NNG, ㅋㅋㅋㅋㅋㅋ_UNKNO...   
172039    True  [사이트_NNG, 망한_VV+ETM, 거_NNB, 인정_NNG, 해라_XSV+EC,...   
17204

In [3]:
used_model = ""

### bag of words

In [4]:
from sklearn.feature_extraction.text import CountVectorizer

def make_bag_of_words(labeled_train, max_features, col_name):
    global used_model
    used_model = "bow"
    
    vectorizer = CountVectorizer(analyzer = "word", tokenizer = None, preprocessor = None,
                                stop_words = None, max_features=max_features)
    
    train_data_features = vectorizer.fit_transform(labeled_train[col_name]).toarray()

    col = ["bow_%s_%s" % (col_name, data) for data in vectorizer.get_feature_names()]
    df_bow = pd.DataFrame(train_data_features, columns = col, index=labeled_train.index)
    
    labeled_train = pd.concat([labeled_train, df_bow],axis=1)
    
    return labeled_train

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/utils/fixes.py:64: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  if 'order' in inspect.getargspec(np.copy)[0]:


## LDA

In [5]:
import gensim
from gensim import corpora, models
from gensim.corpora import TextCorpus, MmCorpus, Dictionary
from nltk.corpus import stopwords
from optparse import OptionParser

def make_lda(train, keep_n, num_topics, col_name):
    global used_model
    used_model = "lda"

    data = train[col_name]

    dictionary = corpora.Dictionary(data)
    dictionary.filter_extremes(keep_n=keep_n)

    corpus = [dictionary.doc2bow(text) for text in data]

    print("Make Lda..")

    lda = gensim.models.ldamodel.LdaModel(corpus=corpus, id2word=dictionary, num_topics=num_topics, chunksize=1000, passes=1)
    
    num = len(train)    
    df = []
    
    for i in range(0,num):
        if i % 10000 == 0:
            print(i)
        
        temp = [i[1] for i in lda.get_document_topics(corpus[i],minimum_probability=0)]
        df.append(temp)
    
    col = ["lda_%s_%d" % (col_name, data) for data in range(0, num_topics)]
    df = pd.DataFrame(df, columns = col)
    df.index = train.index
    
    train = pd.concat([train, df], axis=1)
        
    return train

## TF-IDF

In [6]:
from sklearn.feature_extraction.text import TfidfVectorizer

def make_tf_idf(train, max_features, col_name):
    global used_model
    used_model = "tf-idf"
    
    vectorizer = TfidfVectorizer(min_df = 1, max_features = max_features)
    
    train_data_features = vectorizer.fit_transform(train[col_name]).toarray()
    
    col = ["tfidf_%s_%s" % (col_name, data) for data in vectorizer.get_feature_names()]
    df_tfidf = pd.DataFrame(train_data_features, columns = col, index=train.index)
    
    train = pd.concat([train, df_tfidf],axis=1)
    
    return train

In [7]:
max_features = 1000
#%time labeled_train = make_lda(labeled_train, 5000, 1000, "title_pos")
#%time labeled_train = make_bag_of_words(labeled_train, 300, "title_pos_sentences")
labeled_train = make_tf_idf(labeled_train, max_features, "title_pos_sentences")

### predictor, model 생성

In [8]:
from sklearn.ensemble import RandomForestClassifier

label = 'istroll'
pre = labeled_train.columns.drop(['title_pos', 'title_pos_sentences', label])

model = RandomForestClassifier(n_estimators=10, n_jobs=-1)

### cross validation

In [9]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, labeled_train[pre], labeled_train[label], cv=cv_value, scoring="roc_auc")
cv_result = scores.mean()

print(cv_result)

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.f

0.648875263729


/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/base.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  args, varargs, kw, default = inspect.getargspec(init)
/Library/Frameworks/Python.f

## Result
  * title의 morphs를 bag of words로 변환 (feature : 1,000개) - 0.612752193303
  * title의 pos를 bag of words로 변환 (feature : 1,000개) - 0.630053006037
  * 데이터를 셔플하고 title의 pos를 bag of words로 변환 (feature : 1,000개) - 0.658894903487
  * 데이터를 셔플하고 title의 pos를 tf-idf로 변환 (feature : 1,000개) - 0.648875263729
  
  * title의 pos를 LDA로 변환 (keep_n : 5,000, num_topics : 1,000) - 0.592446285012

In [10]:
import os
import pickle

if not os.path.exists(used_model):
    os.makedirs(used_model)

pickle.dump(labeled_train[pre], open("%s/title_%d.p" % (used_model, max_features), "wb"), protocol = pickle.HIGHEST_PROTOCOL)

istroll                                          title_pos  \
0        False  [명절_NNG, 이_VCP, 지만_EC, 참_VV, 을_ETM, 수_NNB, 없_V...   
1        False  [이젠_NNG+JX, 아니_VCN, 라는_ETM, 증거_NNG, 를_JKO, 대봐_...   
2        False      [뻘_NNG, 글_NNG, +_SY, 오늘_NNG, 의_JKG, 일러스트_NNG]   
3        False                 [일_NR, 일_NNBC, 단식_NNG, 보_VV, 고_EC]   
4        False                            [간헐_NNG, 적_XSN, 단식_NNG]   
5        False                      [옥상_NNG, 에서_JKB, ㅇㅅㅇ_UNKNOWN]   
6        False  [강아지_NNG, 정수기_NNG, 를_JKO, 알아보_VV, 고_EC, 있_VX, ...   
7        False                  [어_IC, 그_NP, 로_JKB, 성공_NNG, !_SF]   
8        False                             [웹툰_NNP, -_SY, 터널_NNG]   
9        False               [혜리_NNG, 의_JKG, 질투_NNG, ㅋㅋㅋ_UNKNOWN]   
10       False  [[_SSO, 꾸준_XR, 글_NNG, ]_SSC, 안녕_NNG, 하_XSV, 세요...   
11       False         [으앙_MAG, 밤_NNG, 이_VCP, 군요_EF, ㅇㅅㅇ_UNKNOWN]   
12       False  [벌레_NNG, 들_XSN, 지_NNG, 들_XSN, 폐쇄_NNG, 서명_NNG, ...   
13       False  [여기_NP, 서_JKB, 농_NNG, 닭_NNG, 체_NNB, 쓰_VV, 다가_E...   
14       False  [이거_NP, 왜_MAG, 농약_NNG, 글_NNG, 이_VCP, 에요_EF, ?_SF]   
15       False  [2_SN, 년_NNB, 째_XSN, 명절_NNG, 에_JKB, 고향_NNG, 에_...   
16       False  [여왕_NNG, 폐하_NNG, 가_JKS, 일_NR, 베_NNG, 유저_NNG, 에...   
17       False  [벌레_NNG, 들_XSN, 이_JKS, 어디_NP, 에서_JKB, 사_VV, 는지...   
18       False                       [신기_NNG, 방기_NNG, ._SF, ._SF]   
19       False                [맥주_NNG, 두_MM, 캔_NNBC, 째_XSN, ._SF]   
20       False  [소고기_NNG, 는_JX, 미국_NNP, 산_XSN, 소고기_NNG, 가_JKS,...   
21       False                 [Free_SL, tempo_SL, -_SY, rain_SL]   
22       False  [저번_NNG, 에_JKB, 광화문_NNG, 에서_JKB, 햄버거_NNG, 먹_VV...   
23       False  [[_SSO, 펌_NNG, ]_SSC, 약자_NNG, 가_JKS, 약자_NNG, 를...   
24       False  [새_MM, 누리_NNG, 의_JKG, '_SY, 특검_NNG, 한계_NNG, 인정...   
25       False     [오늘_MAG, 귀농_NNG, 했_XSV+EP, 습니_EC, 디_NNG, !_SF]   
26       False                      [얼굴_NNG, 만_JX, 반쪽_NNG, 이_JKS]   
27       False  [리버_NNP, 오피스_NNP, 실무_NNG, 에서_JKB, 사용_NNG, 중_NN...   
28       False  [세월_NNG, 호_NNBC, 충돌_NNG, 의혹_NNG, 과_JC, 레이더_NNG...   
29       False          [강원도_NNP, ._SF, .._SY, 흠_IC, ._SF, .._SY]   
...        ...                                                ...   
172014   False                             [배고프_VA, 다아아_EF, 앍_VV]   
172015    True  [CNN_SL, ,_SC, 무인_NNG, 항공기_NNG, 장난감_NNG, 가_VV,...   
172016   False  [윤태호_NNP, 작가_NNP, 님_XSN, 의_JKG, 내부자_NNG, 들_XSN...   
172017   False     [정권_NNG, 별_XSN, 군사비_NNG, 비교_NNG, ._SY, jpg_SL]   
172018   False                   [그거_NP, 생각난다_VV+EC, 농_NNG, ._SF]   
172019    True  [농민_NNG, 님_XSN, 들_XSN, 새로_MAG, 가입_NNG, 했_XSV+E...   
172020   False  [밀_VV+ETM, 덕_NNG, 이_VCP, 라면_EC, 알_VV, 겠_EP, 지만...   
172021    True  [왜_MAG, 솜방망이_NNG, 처벌_NNG, 이_JKS, 계속_MAG, 되_VV,...   
172022   False  [뱃살_NNG, 이_JKS, 찌_VV, 니까_EC, 한_MM, 의원_NNG, 광고_...   
172023   False         [일간_NNG, 워스트_NNP, 시작_NNG, 했_XSV+EP, 닭_NNG]   
172024   False  [동성애자_NNG, 말_VV, 고_EC, 다른_MM, 성소_NNG, 수자_NNG, ...   
172025    True  [김새론_NNP, 10_SN, 몇_NR, 년_NNB, 사이_NNG, 에_JKB, 많...   
172026   False          [다_MAG, 끄_VV, 네_EF, 식_NNB, 대화법_NNG, ._SF]   
172027   False                                        [악동뮤지션_NNP]   
172028   False  [이것_NP, 도_JX, 엄밀히_MAG, 말_NNG, 하_XSV, 면_EC, 광고_...   
172029   False  [[_SSO, 일_NNG, 베_VV+EC, 스_VV, 샷_EP+EP, ]_SSC, ...   
172030   False  [부부_NNG, 의_JKG, 대화_NNG, (_SSO, 말꼬리_NNG, 에_JKB,...   
172031   False  [오랜만_NNG, 에_JKB, 와_VV+EC, 봤_VX+EP, 는데_EC, 잡담_N...   
172032   False  [남자_NNG, 들_XSN, 이_JKS, 버스_NNG, 탔_VV+EP, 을_ETM,...   
172033   False                      [술_NNG, 꾼_XSN, 의_JKG, 결심_NNG]   
172034   False  [닝_NNP, 겐_JKB+JX, ,_SC, 내_NP, 가_JKS, 시범_NNG, 을...   
172035   False  [동성애자_NNG, 에_JKB, 대한_VV+ETM, 박_NNP, 대통령_NNG, 의...   
172036    True  [역시_MAJ, 일_NR, 워_NNG, 가_JKS, 망할_VV+ETM, 줄_NNB,...   
172037    True                        [유익_XR, 한_XSA+ETM, 사이트_NNG]   
172038   False  [이게_NP+JKS, 뭔_MM, 개소리_NNG, 농_NNG, ㅋㅋㅋㅋㅋㅋ_UNKNO...   
172039    True  [사이트_NNG, 망한_VV+ETM, 거_NNB, 인정_NNG, 해라_XSV+EC,...   
17204